tools/driver/emit_pcm_main.cpp

```diff
@@ -0,0 +1,212 @@
+      }
+      return true;
+    }
+
+    if (ParsedArgs.getLastArg(OPT_help)) {
+      std::string Usage = llvm::sys::path::stem(MainExecutablePath);
+      Usage += " -emit-pcm";
+      Table->PrintHelp(llvm::outs(), Usage.c_str(),
+                       "Precompile explicit Clang modules for Swift",
+                       EmitPCMOption, 0, /*ShowAllAliases*/false);
+      return true;
+    }
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_module_name)) {
+      ModuleName = A->getValue();
+      // TODO: Check that the module name is a valid identifier and not "Swift".
+    } else {
+      Diags.diagnose(SourceLoc(), diag::error_no_module_name_given);
+      return true;
+    }
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_sdk))
+      SDKPath = A->getValue();
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_resource_dir))
+      RuntimeResourcePath = A->getValue();
+
+    if (const Arg *A = ParsedArgs.getLastArg(options::OPT_target))
+      Target = llvm::Triple(llvm::Triple::normalize(A->getValue()));
+    else
+      Target = llvm::Triple(llvm::sys::getDefaultTargetTriple());
+
+    for (const Arg *A : ParsedArgs.filtered(OPT_Xcc)) {
+      ClangImporterOpts.ExtraArgs.push_back(A->getValue());
+    }
+
+    std::vector<std::string> InputFilenames;
+    for (const Arg *A : ParsedArgs.filtered(OPT_INPUT)) {
+      InputFilenames.push_back(A->getValue());
+    }
+    if (InputFilenames.empty()) {
+      Diags.diagnose(SourceLoc(), diag::error_mode_requires_an_input_file);
+      return true;
+    }
+    if (InputFilenames.size() > 1) {
+      Diags.diagnose(SourceLoc(), diag::error_mode_requires_one_input_file);
+      return true;
+    }
+    ModuleMapFilename = InputFilenames[0];
+
+    if (const Arg *A = ParsedArgs.getLastArg(OPT_o)) {
+      OutputFilename = A->getValue();
+    } else {
+      // Default to `<module name>.pcm`.
+      llvm::SmallString<256> Path{ModuleName.c_str()};
+      llvm::sys::path::replace_extension(
+          Path, file_types::getExtension(file_types::TY_ClangModuleFile));
+      OutputFilename = Path.str();
+    }
+
+    return false;
+  }
+};
+
+int emit_pcm_main(ArrayRef<const char *> Args, const char *Argv0,
+                  void *MainAddr) {
+  INITIALIZE_LLVM();
+
+  CompilerInstance Instance;
+  PrintingDiagnosticConsumer PDC;
+  if (llvm::sys::Process::StandardErrHasColors())
+    PDC.forceColors();
+  Instance.addDiagnosticConsumer(&PDC);
+
+  EmitPCMInvocation Invocation;
+  Invocation.ClangImporterOpts.Mode =
+      ClangImporterOptions::Modes::PrecompiledModule;
+
+  std::string MainExecutablePath =
+      llvm::sys::fs::getMainExecutable(Argv0, MainAddr);
+  Invocation.MainExecutablePath = MainExecutablePath;
+
+  // Parse arguments.
+  if (Invocation.parseArgs(Args, Instance.getDiags())) {
+    return Instance.getDiags().hadAnyError();
+  }
+
+  StringRef Filename = Invocation.ModuleMapFilename;
+  auto ErrOrBuf = llvm::MemoryBuffer::getFile(Filename);
```

Comment:
This should be going through the virtual file system rather than poking directly at the file system?

Reply:
In this case, probably. I copied this from the modulewrap main, but that might be a different situation since the input file to that action is always an output from a previous action? (Or maybe that should also be updated at some point?)

If I migrate this to a frontend action, then this separate check can probably go away and be folded into something more appropriate.

Reply:
I think modulewrap main should also be updated at some point; it's part of the reason that I think it's useful to make this a frontend action, because it's easier to get this stuff set up properly.